src/main/java/com/google/cloud/spanner/jdbc/JdbcTypeConverter.java

```diff
@@ -183,9 +190,39 @@ static Object convert(Object value, Type type, Class<?> targetType) throws SQLEx
               ZoneId.systemDefault());
         }
       }
+      if (AbstractMessage.class.isAssignableFrom(targetType)) {
+        if (type.getCode() == Code.PROTO || type.getCode() == Code.BYTES) {
+          Method parseMethodParseFrom = targetType.getMethod("parseFrom", byte[].class);
+          return targetType.cast(parseMethodParseFrom.invoke(null, value));
+        }
+      }
+      if (ProtocolMessageEnum.class.isAssignableFrom(targetType)) {
+        if (type.getCode() == Code.ENUM || type.getCode() == Code.INT64) {
+          Method parseMethodForNumber = targetType.getMethod("forNumber", int.class);
+          return targetType.cast(
+              parseMethodForNumber.invoke(
+                  null, AbstractJdbcWrapper.checkedCastToInt((Long) value)));
+        }
+      }
       if (targetType.equals(java.sql.Array.class)) {
         if (type.getCode() == Code.ARRAY) return value;
       }
+      if (targetType.isArray() && type.getCode() == Code.ARRAY) {
+        if (type.getArrayElementType().getCode() == Code.PROTO
+            || type.getArrayElementType().getCode() == Code.BYTES) {
+          Object res = convertArrayOfProtoMessage(value, targetType);
+          if (res != null) {
+            return res;
+          }
+        }
+        if (type.getArrayElementType().getCode() == Code.ENUM
+            || type.getArrayElementType().getCode() == Code.INT64) {
```

Comment:
Same here: Won't this unintentionally try to convert an array of INT64 into an array of PROTO ENUM?